### PR TITLE
feat: config manager to speedup config generation

### PR DESCRIPTION
**Pull request by barak007**
Loading config is slow. we need to eval the config. in dev time this causes a significant delay when refreshing the app. this PR introduce a config manager that can build an optimized version on the config and then load it and watch it.

### CODE REVIEW

File: packages/engine-cli/src/node-config-manager.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    currentValue: Promise<unknown[]> | unknown[];
+    dispose(): Promise<void>;
+    build?: ReturnType<typeof deferred<unknown[]>>;
+};
+
+export class NodeConfigManager {
+    constructor(private config: esbuild.BuildOptions) {}
+    runningBuilds = new Map<string, BuildStats>();
+    hashConfig(entryPoints: string[]) {
+        return hash(entryPoints.join(','));
```
Comment (idoros) [inline packages/engine-cli/src/node-config-manager.ts:16]: sort?

File: packages/engine-cli/src/node-config-manager.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+            const ctx = await this.createBuildTask(entryPoints, {
+                onStart() {
+                    const newBuild = deferred<unknown[]>();
+                    buildStats.error = undefined;
+                    buildStats.currentValue = newBuild.promise;
+                    buildStats.build = newBuild;
+                },
+                onEnd(files) {
+                    buildStats.error = undefined;
```
Comment (idoros) [inline packages/engine-cli/src/node-config-manager.ts:40]: is there a case that build has an error and we still get to build end?

File: packages/engine-cli/src/node-config-manager.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    async dispose() {
+        await this.disposeAll();
+    }
+    private async createBuildTask(
+        entryPoints: string[],
+        hooks: {
+            onStart(): void;
+            onEnd(files: undefined[]): void;
+            onError(err: Error): void;
```
Comment (idoros) [inline packages/engine-cli/src/node-config-manager.ts:84]: connect `onError`

File: packages/engine-cli/src/node-config-manager.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            format: 'cjs',
+            bundle: true,
+            write: false,
+        });
+
+        await Promise.all([ctx.watch(), deferredStart.promise]);
+        return ctx;
+    }
+}
+function deferred<T>() {
```
Comment (idoros) [inline packages/engine-cli/src/node-config-manager.ts:128]: bring from `promise-assist`

File: packages/engine-cli/src/node-config-manager.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        entryPoints: string[],
+        hooks: {
+            onStart(): void;
+            onEnd(files: undefined[]): void;
+            onError(err: Error): void;
+        },
+    ) {
+        const deferredStart = deferred<void>();
+        const ctx = await esbuild.context({
+            ...this.config,
```
Comment (idoros) [inline packages/engine-cli/src/node-config-manager.ts:89]: maybe have tree-shaking by default

File: packages/engine-cli/src/node-config-manager.ts (hunk 1 above)
Comment (barak007) [reply packages/engine-cli/src/node-config-manager.ts:16]: nope. I don't think it's needed in this case. we pass a list of same config from the build so no variance.

File: packages/engine-cli/src/node-config-manager.ts (hunk 5 above)
Comment (barak007) [reply packages/engine-cli/src/node-config-manager.ts:90]: activated

File: packages/engine-cli/src/node-config-manager.ts (hunk 4 above)
Comment (barak007) [reply packages/engine-cli/src/node-config-manager.ts]: done

File: packages/engine-cli/src/node-config-manager.ts (hunk 3 above)
Comment (barak007) [reply packages/engine-cli/src/node-config-manager.ts]: connected

File: packages/engine-cli/src/node-config-manager.ts (hunk 2 above)
Comment (barak007) [reply packages/engine-cli/src/node-config-manager.ts:41]: nope.

File: packages/engine-cli/src/node-config-manager.ts (hunk 1 above)
Comment (barak007) [reply packages/engine-cli/src/node-config-manager.ts]: no sorting the order can will actually matter. but not for us.

**Review by idoros — APPROVED** [review]

**Pull request merged**